dbms/src/Storages/StorageDeltaMerge.cpp

```diff
@@ -839,7 +839,22 @@ DM::PushDownFilterPtr StorageDeltaMerge::buildPushDownFilter(const RSOperatorPtr
         auto [before_where, filter_column_name, _] = ::DB::buildPushDownFilter(pushed_down_filters, *analyzer);
         LOG_DEBUG(tracing_logger, "Push down filter: {}", before_where->dumpActions());
 
-        return std::make_shared<PushDownFilter>(rs_operator, before_where, filter_columns, filter_column_name, extra_cast);
+        auto columns_after_cast = std::make_shared<ColumnDefines>();
+        if (extra_cast != nullptr)
+        {
+            columns_after_cast->reserve(columns_to_read.size());
+            const auto & source_columns = analyzer->getCurrentInputColumns();
+            for (size_t i = 0; i < table_scan_column_info.size(); ++i)
+            {
+                if (table_scan_column_info[i].hasGeneratedColumnFlag())
+                    continue;
+                auto it = columns_to_read_map.at(table_scan_column_info[i].id);
+                RUNTIME_CHECK(it.name == source_columns[i].name);
+                columns_after_cast->push_back(it);
+                columns_after_cast->back().type = source_columns[i].type;
+            }
```

Comment:
1. Is `columns_after_cast` a better name than `source_columns`?
2. is `column` a better name than `it`?
3. Add necessary message: column id, left column name, right column name by `RUNTIME_CHECK_MSG`

Reply:
fixed